docs/docs/custom_preconditioners.md

```diff
@@ -0,0 +1,142 @@
+
+# Defining custom gradient preconditioners
+
+NetKet calls _gradient preconditioner_ that class of techniques that transform the
+gradient of the cost function in such a way to improve convergence properties before
+passing it to an optimiser like :ref:`Sgd` or :ref:`Adam`.
+
+Examples of _gradient preconditioners_ are the [Stochastic Reconfiguration](https://www.attaccalite.com/PhDThesis/html/node15.html) method  (also known as [Natural Gradient](https://wiseodd.github.io/techblog/2018/03/14/natural-gradient/), the Linear Method or second order Hessian-based optimisation.
+
+We call those methods _gradient preconditioners_ because they take as input the gradient of the cost function (usually the energy) and some additional information, and output a transformed gradient.
+
+NetKet comes out of the box with Stocastic Reconfiguration (:ref:`netket.optimiser.SR`), but it is possible to define your own method. 
+If you follow the API outlined in this document, you will be able to use your own preconditioner within netket optimisation drivers without issues.
+
+Keep in mind that writing your own optimisation loop only requires writing about 10  lines of code, and you are not forced to use NetKet's drivers!
+We believe our design to be fairly modular and flexible, and it should be able to accomodate many use-cases, but there will be algorithms that are hard to experess within the boundaries of our API. 
+Do not turn away from netket in those cases, but take all the pieces that you need and write your own optimisation loop!
+
+## The preconditioner interface
+
+Preconditioners are split into two objects for ease of use: 
+ - an `object` which holds all informations necessary to apply the preconditioner, and is constructed from a variational state; 
+ - a `function` taking as input the object above and the gradient of the cost function, and computing the resulting gradient.
```

Comment:
I am sorry, but I have to object to calling the entries of the preconditioner `object` and `function`. These are probably the most generic names possible. (Not to mention that `function` is also an object and is actually called `solver` in the code at the moment.)

So, I understand the wish to be generic, so I guess that's why you don't want to use "quantum Fisher matrix" (which I still prefer to "QGT, tbh) or similar?

But in any case, if you want this to be at the level of abstraction where the preconditioner no longer even has to come from solving a linear system `Ax = b`, the splitting into matrix and solver (or "object" and "function") becomes inapplicable (which is probably what lead to those super generic names) and we'd be better off just letting users pass a preconditioner function taking the state and bare gradient and then providing a `LinearSystemPreconditioner` actually consisting of `lhs_matrix` and `solver`.

Reply:
Because this can be used to do other things than SR.
Like we where cahtting with people wanting to implement the linear method (that's partly why i also wrote those docs).
They won't have the QSR, but another `object`. Also if you do Hessian, you don't really have a QSR.
I know the terminology is horrible.

And also thjey don't really solve a lnear system, but have a function doing some other logic

Reply:
> and we'd be better off just letting users pass a preconditioner function taking the state and bare gradient and then providing a LinearSystemPreconditioner actually consisting of lhs_matrix and solver.

yes, but right now for how this is implemented we need to have both. The function could be nothing, but we still need it for API consistency.
Maybe we hsould make this interface a bit more general..

Reply:
So I see two possible design choices here:

1. We only consider preconditioners that solve a linear system, in which case something like `matrix` (still a bit generic, but in that context probably okay) and `solver` make sense.
2. We want something totally generic (any transformation of the gradient), in which case we should probably not do the splitting between `function` and `object` at the top level but only in a `LinearPreconditioner` subclass.

Reply:
I kind of like this.
So in the end the preconditioner is just a function with signature 
```python
def preconditioner(vstate: VariationalState, grad: PyTree) -> PyTree:
```
And we can have our split for linearpreconditioners.
The only thing we lose is theability to do 
```python
nk.VMC(..., preconditioner=(qgt, solver))
```
and we now need to do 
```python
nk.vmc(...precondtioner=nk.optimizer.LinearPreconditioner(qgt, solver)))
```
but it might even make the intent more clear.

I'd be in favour of htis.

Reply:
We could even hide `preconditioner_restart` inside of `linearpreconditioner`!

Reply:
Yes, that looks pretty good to me! (We could special-case the `perconditioner=<tuple>` case to mean `LinearPreconditioner`; as always, the its a question of how much interface we want to hide for the sake of convenience.)

Reply:
do we have a better name for `lhs_matrix`?
it's actualyl a constructor, not a matrix.

Reply:
> do we have a better name for `lhs_matrix`?

I don't have a better idea; it clearly beats `matrix` and `object, so I am happy with it (also with it being a constructor, it's enough if that is written in the docs, IMHO.)